libcontainer/cgroups/systemd/dbus.go

```diff
@@ -0,0 +1,66 @@
+package systemd
+
+import (
+	"sync"
+
+	systemdDbus "github.com/coreos/go-systemd/v22/dbus"
+)
+
+type dbusConnManager struct {
+	conn     *systemdDbus.Conn
+	rootless bool
+	mut      sync.RWMutex
+}
+
+// newDbusConn initializes systemd dbus connection
+func newDbusConnManager(rootless bool) *dbusConnManager {
+	return &dbusConnManager{
+		rootless: rootless,
+	}
+}
+
+// getDbusConnection lazy initializes systemd dbus connection and returns it
+func (d *dbusConnManager) getDbusConnection() (*systemdDbus.Conn, error) {
+	// In the case where d.conn != nil
+	// Use the read lock the first time to ensure
+	// that Conn can be acquired at the same time.
+	d.mut.RLock()
+	if conn := d.conn; conn != nil {
+		d.mut.RUnlock()
+		return conn, nil
+	}
+	d.mut.RUnlock()
+
+	// In the case where d.conn == nil
+	// Use write lock to ensure that only one
+	// will be created
+	d.mut.Lock()
+	defer d.mut.Unlock()
+	if conn := d.conn; conn != nil {
+		return conn, nil
+	}
+
+	conn, err := d.newDbusConn()
+	if err != nil {
+		return nil, err
+	}
+	d.conn = conn
+	return conn, nil
+}
+
+func (d *dbusConnManager) newDbusConn() (*systemdDbus.Conn, error) {
+	if d.rootless {
+		return NewUserSystemdDbus()
+	}
+	return systemdDbus.New()
+}
+
+// resetDbusConn reset to initial state
+func (d *dbusConnManager) resetDbusConn() {
```

Comment:
I don't see this method used anywhere. Perhaps it should be introduced in the next commit?